injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddAnnotation.java

```diff
@@ -22,21 +22,56 @@
 
 package edu.ucr.cs.riple.injector.changes;
 
-import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.ast.NodeList;
+import com.github.javaparser.ast.expr.AnnotationExpr;
+import com.github.javaparser.ast.expr.MarkerAnnotationExpr;
+import com.github.javaparser.ast.expr.Name;
+import com.github.javaparser.ast.expr.SingleMemberAnnotationExpr;
+import com.github.javaparser.ast.expr.StringLiteralExpr;
+import com.github.javaparser.ast.nodeTypes.NodeWithAnnotations;
 import edu.ucr.cs.riple.injector.location.Location;
+import javax.annotation.Nullable;
 import org.json.simple.JSONObject;
 
 public class AddAnnotation extends Change {
-  public AddAnnotation(Location location, String annotation) {
+
+  /** Argument of the annotation. If null, the added annotation will be a marker annotation. */
+  @Nullable private final String argument;
+
+  public AddAnnotation(Location location, String annotation, @Nullable String argument) {
     super(location, annotation);
+    this.argument = argument;
+  }
+
+  public AddAnnotation(Location location, String annotation) {
+    this(location, annotation, null);
   }
 
   @Override
-  public boolean apply(CompilationUnit tree) {
-    return this.location.apply(tree, annotation, true);
+  public void visit(NodeWithAnnotations<?> node) {
+    NodeList<AnnotationExpr> annotations = node.getAnnotations();
+    AnnotationExpr annotationExpr =
+        this.argument == null
+            ? new MarkerAnnotationExpr(annotationSimpleName)
+            : new SingleMemberAnnotationExpr(
+                new Name(annotationSimpleName), new StringLiteralExpr(argument));
+
+    // Check if annot already exists.
+    boolean annotAlreadyExists =
+        annotations.stream().anyMatch(annot -> annot.equals(annotationExpr));
+    if (annotAlreadyExists) {
+      return;
+    }
+
+    if (argument == null || argument.equals("")) {
+      node.addMarkerAnnotation(annotationSimpleName);
+    } else {
+      node.addAnnotation(annotationExpr);
+    }
   }
 
   @Override
+  @SuppressWarnings("unchecked")
```

Comment:
That annotation was for the whole class, but since I was modifying the class, I thought maybe it is better to refine this and not have it for the full class.

Reply:
Um, are you sure? I don't see the suppression in the "before" code, either at the method or class level...

Reply:
Oh yes, sorry. This one wasn't there before, I just added it to silent the constant IDE warning.